server/shared/src/test/scala/org/http4s/server/middleware/ThrottleSuite.scala

```diff
@@ -27,107 +28,136 @@ import org.http4s.Request
-    }
 
-    takeExtraToken
-      .map { result =>
-        ctx.advanceAndTick(300.milliseconds)
-        result
+      TestControl.execute(takeExtraToken).flatMap { control =>
+        for {
+          _ <- control.results.assertEquals(None)
+          _ <- control.tick
+          _ <- control.results.assertEquals(None)
+          interval <- control.nextInterval
+          _ <- control.advanceAndTick(interval)
+          _ <- control.results
+            .map(_.exists {
+              case Outcome.Succeeded(TokenUnavailable(Some(_))) => true
+              case _ => false
+            })
+            .assert
+        } yield ()
       }
-      .map(_.isInstanceOf[TokenUnavailable])
-      .assert
+    }
   }
 
   test(
     "LocalTokenBucket should only return a single token when only one token available and there are multiple concurrent requests"
   ) {
     val capacity = 1
     val createBucket =
-      TokenBucket.local[IO](capacity, 100.milliseconds)
+      TokenBucket.local[IO](capacity, 1.milliseconds)
 
     val takeTokensSimultaneously = createBucket.flatMap { testee =>
       (1 to 5).toList.parTraverse(_ => testee.takeToken)
     }
 
-    takeTokensSimultaneously
+    val prog = takeTokensSimultaneously
       .map { result =>
         result.count(_ == TokenAvailable)
       }
-      .assertEquals(1)
+
+    TestControl.executeEmbed(prog).assertEquals(1)
   }
 
+  val localGen = for {
+    fd1 <- genFiniteDuration
+    fd2 <- genFiniteDuration.suchThat(_ < fd1)
+  } yield fd1 -> fd2
+
   test(
-    "LocalTokenBucket should return the time until the next token is available when no token is available".flaky
+    "LocalTokenBucket should return the time until the next token is available when no token is available"
   ) {
-    val ctx = TestContext()
-    val capacity = 1
-    val createBucket =
-      TokenBucket.local[IO](capacity, 100.milliseconds)
+    forAllF(localGen) { case (fd1, fd2) =>
+      val capacity = 1
+      val createBucket =
+        TokenBucket.local[IO](capacity, fd1)
 
-    val takeTwoTokens = createBucket.flatMap { testee =>
-      testee.takeToken *> IO.sleep(75.milliseconds) *> testee.takeToken
-    }
+      val takeTwoTokens = createBucket.flatMap { testee =>
+        testee.takeToken *> IO.sleep(fd2) *> testee.takeToken
+      }
 
-    takeTwoTokens.map { result =>
-      ctx.advanceAndTick(75.milliseconds)
-      result match {
-        case TokenUnavailable(t) => t.exists(_ <= 25.milliseconds)
-        case _ => false
+      TestControl.execute(takeTwoTokens).flatMap { control =>
+        for {
+          _ <- control.results.assertEquals(None)
+          _ <- control.tick
+          _ <- control.advanceAndTick(fd2)
+          _ <- control.results
+            .map(_.collect { case Outcome.Succeeded(TokenUnavailable(Some(interval))) =>
+              interval
+            })
+            .map(_.exists(_ == (fd1 - fd2)))
+            .assert
+        } yield ()
       }
```

Comment:
Hmm, is it not possible to rewrite this with `executeEmbed`? At a glance it seems plausible :)